### ISSUE TITLE
markAsRead marks readed every previous message

### ISSUE THREAD

**Issue by gamelaster** [Schmavery/facebook-chat-api#448]
By using the example code from [listen documentation](https://github.com/Schmavery/facebook-chat-api/blob/master/DOCS.md#listen), every previous message is marked as readed.
![image](https://cloud.githubusercontent.com/assets/1665373/24587618/2ddeb254-17ba-11e7-881c-33a3ffa07d5b.png)
Have someone same behaviour?
I looked on mark_as_read.php request, but not looks like something has been changed.

**Comment by mangopearapples** [comment]
I think I can reproduce. But for me, it's not the previous message that is marked as read, it's the first message is marked as read and then the `markAsRead` function won't work again.

I tested with this:

```
        api.listen((err, message)=>{
		if(message.type == 'message'){
			console.log('New message');
			api.markAsRead(message.threadID, (err)=>{
				if(err) return console.log(err);
				console.log('Marked as read');
			});
		}
	});
```
Only the first time the api receives a message will it actually mark it as read. It will not mark as read again until I restart the app. This applies to different threads too. In effect, only the first message the api marks as read will actually be marked as read regardless of thread.

**Comment by mangopearapples** [comment]
> I looked on mark_as_read.php request,

@GAMELASTER Not sure if I'm being dumb but where are you getting `mark_as_read.php` from? I'm seeing `change_read_status.php` on facebook and it's also what the api is using.

**Comment by GAMELASTER** [comment]
@mangopearapples sorry, I'm beign dumb! I have `change_read_status.php` too, but by mistake written wrong name 😆

**Comment by GAMELASTER** [comment]
@mangopearapples actually it's have something similar behaviour like you. Just look gif:
![2017-04-02_23-13-30](https://cloud.githubusercontent.com/assets/1665373/24591098/1e0cae04-17fa-11e7-8426-c041b1f8cc27.gif)

**Comment by mangopearapples** [comment]
Actually yeah I get the exact same results as you. I just couldn't see the previous "seen" status on the normal facebook home page.

**Comment by ravkr** [comment]
I haven't seen this bug yet, but I will investigate it tomorrow

**Comment by bobland1** [comment]
If message.body != undefined should fix the issue. You receive undefineds occasionally. Make sure u are marking sure u remove all undefineds.

If message != undefined {
    If message.thread != undefined {
       Api.read ....
    }
}

**Comment by bobland1** [comment]
Once you have filtered out all undefineds test. And report back

**Comment by bobland1** [comment]
Try testing function (err) instead of => it's better and more reliable

**Comment by mangopearapples** [comment]
Interestingly, when the program above is running, even the actual facebook messenger on the website will not mark the latest message as read. However when I view the message in Android's FB messenger, it will then mark the message as read correctly.

**Comment by Schmavery** [comment]
@bobland1 I'm not sure if checking if the message.body is undefined is the right approach.  You should check message.type and err, if that doesn't work, there's a bug in the api 😛 

@mangopearapples I wouldn't be surprised if this could be a problem on facebook's end, but it's nice to investigate anyway :)

**Comment by bobland1** [comment]
@Schmavery I was using it more as an example. I meant for him to filter out the undefineds of message, message.threadID and message.type
```
if message != undefined {
      if message.type != undefined {
          if message.threadID != undefined {
              api.markAsRead(message.threadID, function (err) {
                     if(err) return console.log(err);
	             console.log("Marked as read");
              });
          }
     }
}

```

**Comment by Schmavery** [comment]
@bobland1 my point was that I don't think this is the right solution to this problem, but to each their own.

**Comment by bobland1** [comment]
Oh I understand :) sorry

**Comment by mangopearapples** [comment]
Sometimes messenger POST's to `mark_seen.php` which `markAsRead` doesn't do. Maybe this has something to do with this issue.

The form data for the `mark_seen.php` request is just `seen_timestamp: [same timestamp as change_read_status.php]`

![fb_messenger_form-data](https://cloud.githubusercontent.com/assets/3869445/24659540/e0ceed26-1944-11e7-9522-5b5b7287d654.png)

**Comment by GAMELASTER** [comment]
@mangopearapples lol, so I don't has been blind and there is really a `mark_seen.php` request haha 😆

**Comment by GAMELASTER** [comment]
It's again have another behaviour! Now it's sometimes read until end, sometimes ignoring few messages and then again to end

**Comment by Schmavery** [comment]
😕 mark as read seems to work perfectly on our bots these days... I'm assuming people are working from the latest version? (I just published 1.4.0 btw)

**Comment by bsansouci** [comment]
Closing as it seems like it's working. Please re-open if it's still an issue.
We have 2 bots running on this API that are working fine.
